File: aspn-py/src/aspn23/type_image_feature.py
"""
This code is generated via firehose.
DO NOT hand edit code.  Make any changes required using the firehose repo instead.
"""


from dataclasses import dataclass
from typing import Optional

import numpy as np
from numpy.typing import NDArray


@dataclass
class TypeImageFeature:
    """
    Image feature characteristics

    ### Attributes

    response - float:
            Value representing the magnitude of the response of the keypoint detection algorithm
            used to locate this keypoint in the image.

    orientation - float:
            Value representing the orientation of the detected keypoint, as measured relative to
            the image coordinate system (radians, clockwise). Metadata flag determines if this field
            is calculated by the current keypoint detector.

    size - float:
            Effective size (pixels) of the detected keypoint. This field takes into
            consideration the octave and kernel used to localize the keypoint, and normalizes to the
            original size of the image.

    class_id - int:
            ID that ties the detected keypoint to a class or group of keypoints.

    octave - int:
            Octave or pyramid layer from which the detected keypoint was extracted.

    descriptor - NDArray[np.int64]:
            Feature descriptor.
    """

    response: float
    orientation: float
    size: float
    class_id: int
    octave: int
    descriptor: NDArray[np.int64]

File: aspn-c/src/aspn23/TypeImageFeature.c

/*
 * This code is generated via firehose.
 * DO NOT hand edit code.  Make any changes required using the firehose repo instead
 */

#include "TypeImageFeature.h"

Aspn23TypeImageFeature* ASPN_NULLABLE aspn23_type_image_feature_new(double response,
                                                                    double orientation,
                                                                    double size,
                                                                    uint16_t class_id,
                                                                    uint16_t octave,
                                                                    uint16_t descriptor_size,
                                                                    uint8_t* descriptor) {
	Aspn23TypeImageFeature* self =
	    (struct Aspn23TypeImageFeature*)malloc(sizeof(Aspn23TypeImageFeature));
	if (NULL == self) return NULL;
	self->response        = response;
	self->orientation     = orientation;
	self->size            = size;
	self->class_id        = class_id;
	self->octave          = octave;
	self->descriptor_size = descriptor_size;

	self->descriptor = NULL;
	if (descriptor != NULL && descriptor_size != 0) {
		if (descriptor_size == 0)
			self->descriptor = NULL;
		else if (descriptor_size > 0) {

			self->descriptor = (uint8_t*)calloc(descriptor_size, sizeof(uint8_t));
			memcpy(self->descriptor, descriptor, sizeof(uint8_t) * descriptor_size);

		} else {
			fprintf(
			    stderr,
			    "An error occurred: '%s' defines the length '%s' and must be a positive integer",
			    "descriptor_size",
			    "descriptor");
			aspn23_type_image_feature_free(self);
			return NULL;
		}
	}

	return self;
}

Aspn23TypeImageFeature* ASPN_NULLABLE
aspn23_type_image_feature_copy(Aspn23TypeImageFeature* input) {

	Aspn23TypeImageFeature* out = aspn23_type_image_feature_new(input->response,
	                                                            input->orientation,
	                                                            input->size,
	                                                            input->class_id,
	                                                            input->octave,
	                                                            input->descriptor_size,
	                                                            input->descriptor);

	return out;
}

void aspn23_type_image_feature_free(void* pointer) {
	Aspn23TypeImageFeature* self = (Aspn23TypeImageFeature*)pointer;
	if (NULL == self) return;
	aspn23_type_image_feature_free_members(self);
	free(self);
}

void aspn23_type_image_feature_free_members(Aspn23TypeImageFeature* self) {
	if (NULL == self) return;

	if (self->descriptor != NULL && self->descriptor_size != 0) {
		free(self->descriptor);
	}
}
